From 2af6dd682524a3e98991fa57a35983a9437442b5 Mon Sep 17 00:00:00 2001
From: Zhang Rui <rui.zhang@intel.com>
Date: Sun, 16 Mar 2014 21:34:16 +0800
Subject: [PATCH 03/23] ACPI: add ACPI_TYPE_LOCAL_REFERENCE support to
 acpi_extract_package()

Add ACPI_TYPE_LOCAL_REFERENCE support to acpi_extract_package(),
so that we can use this helper for more cases like _ART/_TRT.

Change-Id: I9cfec60a2e6c5b44c00552aca6dbee8dbe2d0b20
Signed-off-by: Zhang Rui <rui.zhang@intel.com>
---
 drivers/acpi/utils.c |   29 ++++++++++++++++++++++++++++-
 1 file changed, 28 insertions(+), 1 deletion(-)

diff --git a/drivers/acpi/utils.c b/drivers/acpi/utils.c
index 85e3b61..7821275 100644
--- a/drivers/acpi/utils.c
+++ b/drivers/acpi/utils.c
@@ -148,6 +148,21 @@ acpi_extract_package(union acpi_object *package,
 				break;
 			}
 			break;
+		case ACPI_TYPE_LOCAL_REFERENCE:
+			switch (format_string[i]) {
+			case 'R':
+				size_required += sizeof(void *);
+				tail_offset += sizeof(void *);
+				break;
+			default:
+				printk(KERN_WARNING PREFIX "Invalid package element"
+					      " [%d] got reference,"
+					      " expecting [%c]\n",
+					      i, format_string[i]);
+				return AE_BAD_DATA;
+				break;
+			}
+			break;
 
 		case ACPI_TYPE_PACKAGE:
 		default:
@@ -247,7 +262,19 @@ acpi_extract_package(union acpi_object *package,
 				break;
 			}
 			break;
-
+		case ACPI_TYPE_LOCAL_REFERENCE:
+			switch (format_string[i]) {
+			case 'R':
+				*(void **)head =
+				    (void *)element->reference.handle;
+				head += sizeof(void *);
+				break;
+			default:
+				/* Should never get here */
+				break;
+			}
+			break;
+			
 		case ACPI_TYPE_PACKAGE:
 			/* TBD: handle nested packages... */
 		default:
-- 
1.7.9.5

